### PR TITLE
gh-109218: Improve documentation for the complex() constructor

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-109218

----
📚 Documentation preview 📚: https://cpython-previews--119687.org.readthedocs.build/

### CODE REVIEW

**Review by skirpichev — APPROVED** [review]
LGTM

Few remarks:
0) This touch also bool, float and int constructors; changes looks fine, but at least the commit message could be more verbose and mention that.
1) Should be merged before #119620 
2) cmath has yet another instance of ``z == z.real + z.imag*1j`` invariant, mentioned in the issue thread.  I think this should be fixed too, probably just by removing that line.

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -371,29 +372,64 @@ are always available.  They are listed here in alphabetical order.
...
-   Return a complex number with the value *real* + *imag*\*1j or convert a string
-   or number to a complex number.  If the first parameter is a string, it will
-   be interpreted as a complex number and the function must be called without a
-   second parameter.  The second parameter can never be a string. Each argument
-   may be any numeric type (including complex).  If *imag* is omitted, it
-   defaults to zero and the constructor serves as a numeric conversion like
-   :class:`int` and :class:`float`.  If both arguments are omitted, returns
-   ``0j``.
+   Convert a string or a number to a complex number or create a complex number
+   from the real and imaginary parts.
```
Comment (AlexWaygood) [inline Doc/library/functions.rst]:
```suggestion
   Convert a single string or number to a complex number, or create a
   complex number from real and imaginary parts.
```

File: Doc/library/functions.rst [hunk 2]
```diff
@@ -371,29 +372,64 @@ are always available.  They are listed here in alphabetical order.
...
+   preceded by a sign, and optionally followed by the ``j`` or ``J`` suffix,
+   or a pair of decimal numbers, optionally preceded by a sign, separated by
+   the ``+`` or ``-`` operator, and followed by the ``j`` or ``J`` suffix.
+   A string representing a NaN (not-a-number), or infinity is acceptable
+   in place of a decimal number, like in :func:`float`.
+   The string can optionally be surrounded by whitespaces and the round
+   parenthesis ``(`` and ``)``, which are ignored.
+   The string must not contain whitespace between ``+``, ``-``, the ``j`` or
+   ``J`` suffix, and the decimal number.  For example, ``complex('1+2j')`` is fine, but
+   ``complex('1 + 2j')`` raises :exc:`ValueError`.
```
Comment (AlexWaygood) [inline Doc/library/functions.rst]: This is a lot of detail for the second paragraph. Users might find it easier to understand how to create complex numbers if the REPL examples came before this full specification

File: Doc/library/functions.rst [hunk 3]
```diff
@@ -371,29 +372,64 @@ are always available.  They are listed here in alphabetical order.
...
+   If one of arguments is a real number, only its real component is used in
+   the above expressions.
+
+   If all arguments are omitted, returns ``0j``.
 
-      When converting from a string, the string must not contain whitespace
-      around the central ``+`` or ``-`` operator.  For example,
-      ``complex('1+2j')`` is fine, but ``complex('1 + 2j')`` raises
-      :exc:`ValueError`.
+   Examples::
```
Comment (AlexWaygood) [inline Doc/library/functions.rst]:
```suggestion
   Examples:

   .. doctest::
```

File: Doc/library/functions.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst]: I followed the example of `float()` which has even more detailed description of the input format in the second paragraph. I merged it with a note about whitespaces, so all information related to parsing string to complex is now in one place.

I would be happy to make the description shorter, even if less detailed. Do you have any suggestions?

Comment (AlexWaygood) [reply Doc/library/functions.rst]: Fair enough, though I would honestly also be inclined to move the `float()` examples higher up as well. I think it's probably correct to provide a full specification here; I just worry that it could overwhelmingly technical for Python beginners, who will probably learn best by looking at the examples.

In any event, I don't have a strong opinion; I'm happy to defer to you and Mark

File: Doc/library/functions.rst [hunk 4]
```diff
@@ -141,10 +141,11 @@ are always available.  They are listed here in alphabetical order.
    See also :func:`format` for more information.
 
 
-.. class:: bool(x=False)
+.. class:: bool(object=False, /)
```
Comment (mdickinson) [inline Doc/library/functions.rst:144]: Was including this intentional, or was this change intended for a different PR?

File: Doc/library/functions.rst (hunk 2 above)
Comment (mdickinson) [reply Doc/library/functions.rst]: I think @AlexWaygood has a good point here: for readability, it would be nice to see just three examples (one for each of the three signatures) following the initial "Convert a single string or number to a complex number, or ..." line, before we get into the heavy detail of exactly which strings are permitted.

File: Doc/library/functions.rst [hunk 5]
```diff
@@ -371,29 +372,66 @@ are always available.  They are listed here in alphabetical order.
...
-      around the central ``+`` or ``-`` operator.  For example,
-      ``complex('1+2j')`` is fine, but ``complex('1 + 2j')`` raises
-      :exc:`ValueError`.
+   If two arguments are provided or keyword arguments are used, each argument
+   may be any numeric type (including complex).
+   If both arguments are real numbers, return a complex number with the real
+   component *real* and the imaginary component *imag*.
+   If both arguments are complex numbers, return a complex number with the real
+   component ``real.real-imag.imag`` and the imaginary component
+   ``real.imag+imag.real``.
```
Comment (mdickinson) [inline Doc/library/functions.rst:437]: Presumably this will mention the deprecation if/when #119620  is merged? Or is the plan to remove the mention of allowing complex altogether?

I'd be tempted to not even document this (mis)feature, on the basis that people who already know about it already know about it, and people who don't know about it shouldn't start using it.

File: Doc/library/functions.rst [hunk 6]
```diff
@@ -926,17 +965,19 @@ are always available.  They are listed here in alphabetical order.
...
+   Return an integer object constructed from a number or a string, or return
+   ``0`` if no arguments are given.
 
-   Return an integer object constructed from a number or string *x*, or return
-   ``0`` if no arguments are given.  If *x* defines :meth:`~object.__int__`,
-   ``int(x)`` returns ``x.__int__()``.  If *x* defines :meth:`~object.__index__`,
-   it returns ``x.__index__()``.  If *x* defines :meth:`~object.__trunc__`,
+   If the argument defines :meth:`~object.__int__`,
+   ``int(x)`` returns ``x.__int__()``.  If the argument defines :meth:`~object.__index__`,
+   it returns ``x.__index__()``.  If the argument defines :meth:`~object.__trunc__`,
```
Comment (mdickinson) [inline Doc/library/functions.rst:1004]: Darn: we haven't removed the delegation to `__trunc__` yet? Looks like it was deprecated in 3.11, so maybe we can remove it for 3.14?

Off-topic for this PR, of course.

**Review by mdickinson — APPROVED** [review]
LGTM.

This is very much a judgement call and personal opinion, and I don't know how well it fits with the general direction of our documentation, but I do think it would be nice to have a small number of examples immediately following the initial "Convert a single string ..." doc line.

File: Doc/library/functions.rst (hunk 4 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst:144]: Yes, it was intentional. All for types are numeric. I tried to make the `complex()` description similar to `float()` and `int()` descriptions, so I changed them also a little. The `complex()` signature now clearly indicates positional-only parameters, and I used meaningful names for them. But other constructors still used the `x` parameter without indicating that it is positional-only (in the past it _was_ keyword-or-positional), so I updated them too and removed references to the `x` name.

These were the minimal changes to other constructors. They may need more changes, as well as constructors of non-numeric types and other functions, but this is too far out of the scope of this PR.

File: Doc/library/functions.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst]: I have moved examples for `complex()` and `float()` and added examples for `int()`.

File: Doc/library/functions.rst (hunk 5 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst:437]: The plan was to leave this until the end of the deprecation period.

I would be happy to remove this, but then we will receive another reports about wrong documentation or behavior. The old `real + imag*1j` was more correct for input with finite components (and ignoring corner cases with negative zero).

"Each argument may be any numeric type (including complex)." is already in the current documentation, so it is late to hide this without also changing the behavior.

Comment (mdickinson) [reply Doc/library/functions.rst:437]: Fair enough.

File: Doc/library/functions.rst [hunk 7]
```diff
@@ -371,29 +372,66 @@ are always available.  They are listed here in alphabetical order.
...
+      >>> complex('\t( -1.23+4.5J )\n')
+      (-1.23+4.5j)
+      >>> complex('-Infinity+NaNj')
+      (-inf+nanj)
+      >>> complex(1.23)
+      (1.23+0j)
+      >>> complex(imag=-4.5)
+      -4.5j
+      >>> complex(-1.23, 4.5)
+      (-1.23+4.5j)
```
Comment (AlexWaygood) [inline Doc/library/functions.rst:401]: I suggest we reorder some of these so that the stranger ones come below the more common ones:

```suggestion
      >>> complex('+1.23')
      (1.23+0j)
      >>> complex('-4.5j')
      -4.5j
      >>> complex('-1.23+4.5j')
      (-1.23+4.5j)
      >>> complex(-1.23, 4.5)
      (-1.23+4.5j)
      >>> complex(1.23)
      (1.23+0j)
      >>> complex(imag=-4.5)
      -4.5j
      >>> complex('\t( -1.23+4.5J )\n')
      (-1.23+4.5j)
      >>> complex('-Infinity+NaNj')
      (-inf+nanj)
```

**Review by AlexWaygood — APPROVED** [review]

File: Doc/library/functions.rst (hunk 7 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst:401]: But this mixes examples for three fundamentally different roles: string parsing, numerical conversion and construction from components.

File: Doc/library/functions.rst (hunk 5 above)
Comment (skirpichev) [reply Doc/library/functions.rst:437]: Best we can do is to put the deprecation note as close to this sentence, as possible:)

File: Doc/library/functions.rst (hunk 7 above)
Comment (AlexWaygood) [reply Doc/library/functions.rst:401]: It wasn't immediately obvious to me that that was the reason behind your ordering, and I doubt it would be obvious to beginners either, which is why I suggested a different order of "most useful to least useful". But again, I don't have a strong opinion; your PR is certainly fine as-is :-)

File: Doc/library/functions.rst [hunk 8]
```diff
@@ -371,29 +372,73 @@ are always available.  They are listed here in alphabetical order.
...
-.. class:: complex(real=0, imag=0)
-           complex(string)
+.. class:: complex(number=0, /)
+           complex(string, /)
+           complex(real=0, imag=0)
+
+   Convert a single string or number to a complex number, or create a
+   complex number from real and imaginary parts.
+
+   Examples:
```
Comment (picnixz) [inline Doc/library/functions.rst:382]:
```suggestion
   .. rubric:: Examples
```

It may look a bit nicer, but I don't know whether you want a smaller heading here.

File: Doc/library/functions.rst (hunk 7 above)
Comment (picnixz) [reply Doc/library/functions.rst:401]: I would also expect the most useful to the least useful examples even though it mixes the parsers.

File: Doc/library/functions.rst [hunk 9]
```diff
@@ -371,29 +372,73 @@ are always available.  They are listed here in alphabetical order.
...
+      >>> complex(1.23)
+      (1.23+0j)
+      >>> complex(imag=-4.5)
+      -4.5j
+      >>> complex(-1.23, 4.5)
+      (-1.23+4.5j)
+
+   If the argument is a string, it must contain either a real part (in the
+   same format as for :func:`float`) or an imaginary part (in the same
+   format but with a ``'j'`` or ``'J'`` suffix ), or both real and imaginary
```
Comment (picnixz) [inline Doc/library/functions.rst]:
```suggestion
   format but with a ``'j'`` or ``'J'`` suffix), or both real and imaginary
```

File: Doc/library/functions.rst [hunk 10]
```diff
@@ -371,29 +372,73 @@ are always available.  They are listed here in alphabetical order.
...
+   parts (the sign of the imaginary part is mandatory in this case).
+   The string can optionally be surrounded by whitespaces and the round
+   parenthesis ``'('`` and ``')'``, which are ignored.
+   The string must not contain whitespace between ``'+'``, ``'-'``, the
+   ``'j'`` or ``'J'`` suffix, and the decimal number.
+   For example, ``complex('1+2j')`` is fine, but ``complex('1 + 2j')`` raises
+   :exc:`ValueError`.
+   More precisely, the input must conform to the :token:`~float:complexvalue`
+   production rule in the following grammar, after parenthesis and leading and
```
Comment (picnixz) [inline Doc/library/functions.rst]:
```suggestion
   production rule in the following grammar, after parentheses and leading and
```

File: Doc/library/functions.rst [hunk 11]
```diff
@@ -371,29 +372,73 @@ are always available.  They are listed here in alphabetical order.
...
+      (-1.23+4.5j)
+
+   If the argument is a string, it must contain either a real part (in the
+   same format as for :func:`float`) or an imaginary part (in the same
+   format but with a ``'j'`` or ``'J'`` suffix ), or both real and imaginary
+   parts (the sign of the imaginary part is mandatory in this case).
+   The string can optionally be surrounded by whitespaces and the round
+   parenthesis ``'('`` and ``')'``, which are ignored.
+   The string must not contain whitespace between ``'+'``, ``'-'``, the
+   ``'j'`` or ``'J'`` suffix, and the decimal number.
```
Comment (skirpichev) [inline Doc/library/functions.rst:410]: Nit: this "j" vs "J" happens in several places.  You could just mention (as in the float constructor) that parsing is case-insensitive:
```pycon
>>> complex('(InFiNITY+1j)')
(inf+1j)
```

File: Doc/library/functions.rst (hunk 8 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst:382]: Looking how it is used in `Doc/library/asyncio-api-index.rst`, I do not think that it is appropriate here.

File: Doc/library/functions.rst (hunk 11 above)
Comment (serhiy-storchaka) [reply Doc/library/functions.rst:410]: But "e" and "E" are mentioned separately in the float grammar. We can say also that all whitespaces except the leading and trailing are not acceptable, but I think it will be clearer to be a little more verbose here.

File: Doc/library/functions.rst [hunk 12]
```diff
@@ -371,29 +372,73 @@ are always available.  They are listed here in alphabetical order.
...
+      -4.5j
+      >>> complex(-1.23, 4.5)
+      (-1.23+4.5j)
+
+   If the argument is a string, it must contain either a real part (in the
+   same format as for :func:`float`) or an imaginary part (in the same
+   format but with a ``'j'`` or ``'J'`` suffix ), or both real and imaginary
+   parts (the sign of the imaginary part is mandatory in this case).
+   The string can optionally be surrounded by whitespaces and the round
+   parenthesis ``'('`` and ``')'``, which are ignored.
```
Comment (serhiy-storchaka) [inline Doc/library/functions.rst]:
```suggestion
   parentheses ``'('`` and ``')'``, which are ignored.
```

File: Doc/library/functions.rst [hunk 13]
```diff
@@ -682,21 +727,38 @@ are always available.  They are listed here in alphabetical order.
...
 
    If the argument is a string, it should contain a decimal number, optionally
    preceded by a sign, and optionally embedded in whitespace.  The optional
    sign may be ``'+'`` or ``'-'``; a ``'+'`` sign has no effect on the value
    produced.  The argument may also be a string representing a NaN
-   (not-a-number), or positive or negative infinity.  More precisely, the
-   input must conform to the ``floatvalue`` production rule in the following
-   grammar, after leading and trailing whitespace characters are removed:
+   (not-a-number), or positive or negative infinity.
+   More precisely, the input must conform to the :token:`~float:`floatvalue`
```
Comment (serhiy-storchaka) [inline Doc/library/functions.rst]:
```suggestion
   More precisely, the input must conform to the :token:`~float:floatvalue`
```
